springfox-schema/src/main/java/springfox/documentation/schema/JacksonJsonViewProjectionProvider.java

```diff
@@ -0,0 +1,29 @@
+package springfox.documentation.schema;
+
+import java.lang.annotation.Annotation;
+import java.util.List;
+
+import com.fasterxml.classmate.ResolvedType;
+import com.fasterxml.jackson.annotation.JsonView;
+import com.google.common.base.Optional;
+import com.google.common.collect.FluentIterable;
+
+import springfox.documentation.spi.DocumentationType;
+import springfox.documentation.spi.schema.ModelProjectionProviderPlugin;
+
+public class JacksonJsonViewProjectionProvider implements ModelProjectionProviderPlugin {
+
+  @Override
+  public boolean supports(DocumentationType delimiter) {
+    return true;
+  }
+
+  @Override
+  public Optional<Class<?>> projectionFor(ResolvedType type, List<Annotation> typeAnnotations) {
```

Comment:
This should return a `ResolvedType` IMO.

Also have you thought about how we might influence a projection not using annotations (I don't know how gson handles this)? I think it might be a good idea to extract a parameter object i.e. projection context that gives it access to/or parts of the `RequestMappingContext`

Reply:
Oh never mind... I see what this is doing. I don't think this interface is accurate. I'm thinking of the return value as an __alternate type__ for the `type`. This makes this interface is jackson specific and will not work for spring-data-rest I think. 

I was thinking more in these terms. Lets take this example [from here](http://www.baeldung.com/jackson-json-view-annotation)

```java
public class Views {
    public static class Public {
    }
 
    public static class Internal extends Public {
    }
}

public class Item {
  
    @JsonView(Views.Public.class)
    public int id;
 
    @JsonView(Views.Public.class)
    public String itemName;
 
    @JsonView(Views.Internal.class)
    public String ownerName;
}
```

I would think that this interface would behave something like this.

```java
//Pseudo code on how this should be used.

ProjectionContext context = new ProjectionContext(
          type, //Type to find a projection type
          annotations, //Annotations to help finding projections
          ... more params as needed
         );
JacksonJsonViewProjectionProvider plugin = createInstanceOfPlugin(); 
Optional<ResolvedType> projection = instance.projectionFor(context);
//OR
Optional<Model> projection = instance.projectionFor(context);
```
Here projection would be a either a in-memory `ResolvedType` of type `ItemPublicView` built using [AlternateTypeBuilder](https://github.com/springfox/springfox/blob/master/springfox-core/src/main/java/springfox/documentation/builders/AlternateTypeBuilder.java ) or straight away building the Model object via the `ModelBuilder`, that has the public properties as we do here. What do you think?

Reply:
As i know, Gson does not have full projection support. Just only `SerializationStrategy` to determine which properties will be serialized.

Reply:
There is a method `Optional <Class<?> getRequiredAnnotation()` that will handle case with **JsonView**, and will be ignored in case of *Spring Data Rest*

Reply:
And we might have a case< where there won't be any projection, just all Class... For example, no defined `@Projection` classes, or no **@JsonView** annotation on field.

Reply:
Dilip, that is outdated code....

```
 @Override
  public List<ResolvedType> projectionsFor(ResolvedType type, Optional<? extends Annotation> requiredAnnotation) {
    List<Class<?>> projections = newArrayList();
    if (requiredAnnotation.isPresent() &&
        requiredAnnotation.get() instanceof JsonView) {
      projections = newArrayList(((JsonView)requiredAnnotation.get()).value());
    }
    return from(projections).transform(toResolvedType()).toList();
  }
```

Reply:
In last commit I changed interface and implementation

Reply:
I agreed, `ProjectionContext` is a good idea, to consolidate all required information to determine projection

Reply:
But, for me *Projection* is **ONLY** a part of `ModelContext`, just additional information, to determine which properties must be serialized or deserialized.